full-service/src/service/sync.rs

```diff
@@ -143,141 +142,132 @@ pub fn sync_account(
-                        None => return None,
-                        Some(a) => a,
-                    };
-                    let subaddress_index = decode_subaddress_index(
-                        &tx_out,
-                        view_account_key.view_private_key(),
-                        &subaddress_keys,
-                    );
-                    Some((block_index, tx_out, amount, subaddress_index))
-                })
-                .collect();
-            let num_received_txos = received_txos.len();
-
-            // Write received transactions to the database.
-            for (block_index, tx_out, amount, subaddress_index) in received_txos {
-                Txo::create_received(
-                    tx_out.clone(),
-                    subaddress_index,
-                    None,
-                    amount,
-                    block_index,
-                    account_id_hex,
-                    conn,
-                )?;
-            }
+    }
 
-            // Match key images to mark existing unspent transactions as spent.
-            let unspent_key_images: HashMap<KeyImage, String> =
-                Txo::list_unspent_or_pending_key_images(account_id_hex, None, conn)?;
-            let spent_txos: Vec<(u64, String)> = key_images
-                .into_par_iter()
-                .filter_map(|(block_index, key_image)| {
-                    unspent_key_images
-                        .get(&key_image)
-                        .map(|txo_id_hex| (block_index, txo_id_hex.clone()))
-                })
-                .collect();
-            let num_spent_txos = spent_txos.len();
-            for (block_index, txo_id_hex) in &spent_txos {
-                Txo::update_spent_block_index(txo_id_hex, *block_index as u64, conn)?;
-                TransactionLog::update_pending_associated_with_txo_to_succeeded(
-                    txo_id_hex,
-                    *block_index,
-                    conn,
-                )?;
-            }
+    // If no blocks were found, exit.
+    if end_block_index.is_none() {
+        return Ok(SyncStatus::NoMoreBlocks);
+    }
+    let end_block_index = end_block_index.unwrap();
+
+    if account.view_only {
+        let view_account_key: ViewAccountKey = mc_util_serial::decode(&account.account_key)?;
+
+        // Attempt to decode each transaction as received by this account.
+        let received_txos: Vec<_> = tx_outs
+            .into_par_iter()
+            .filter_map(|(block_index, tx_out)| {
+                let amount = match decode_amount(&tx_out, view_account_key.view_private_key()) {
+                    None => return None,
+                    Some(a) => a,
+                };
+                let subaddress_index = decode_subaddress_index(
+                    &tx_out,
+                    view_account_key.view_private_key(),
+                    &wallet_db.get_conn().unwrap(),
```

Comment:
This should not call unwrap, it should propagate the error

Reply:
Thanks for catching, leftover from draft